amazon_transcribe/serialize.py

```diff
@@ -12,75 +12,120 @@
+from amazon_transcribe.request import Request
 from amazon_transcribe.structures import BufferableByteStream
 from amazon_transcribe.utils import _add_required_headers
-from amazon_transcribe.exceptions import SerializerException
+from amazon_transcribe.exceptions import SerializerException, ValidationException
 from amazon_transcribe.model import (
     AudioEvent,
     StartStreamTranscriptionRequest,
     BaseEvent,
 )
 
-HEADER_VALUE = Union[int, None, str]
 
-
-class Serializer:
-    def __init__(self):
-        raise NotImplementedError("Serializer")
-
-    def serialize(self) -> Tuple[Dict[str, HEADER_VALUE], BufferedIOBase]:
-        """Serialize out to payload and headers."""
-        raise NotImplementedError("serialize")
-
-    def serialize_to_request(self) -> PreparedRequest:
-        """Serialize parameters into an HTTP request."""
-        raise NotImplementedError("serialize_to_request")
-
-
-class TranscribeStreamingRequestSerializer(Serializer):
+class TranscribeStreamingSerializer:
     """Convert StartStreamTranscriptionRequest into a
     Request object for streaming to the Transcribe service.
     """
 
-    def __init__(self, endpoint, transcribe_request):
-        self.endpoint: str = endpoint
-        self.method: str = "POST"
-        self.request_uri: str = "/stream-transcription"
-        self.request_shape: StartStreamTranscriptionRequest = transcribe_request
+    def _raise_param_validation_exception(
+        self, value_type: str, header: str, expected_type: str
+    ):
+        raise ValidationException(
+            f"Unexpected param type (value_type) encountered for {header}, expected type: {expected_type}"
+        )
 
-    def serialize(self) -> Tuple[Dict[str, HEADER_VALUE], BufferedIOBase]:
-        headers = {
-            "x-amzn-transcribe-language-code": self.request_shape.language_code,
-            "x-amzn-transcribe-sample-rate": self.request_shape.media_sample_rate_hz,
-            "x-amzn-transcribe-media-encoding": self.request_shape.media_encoding,
-            "x-amzn-transcribe-vocabulary-name": self.request_shape.vocabulary_name,
-            "x-amzn-transcribe-session-id": self.request_shape.session_id,
-            "x-amzn-transcribe-vocabulary-filter-method": self.request_shape.vocab_filter_method,
-            "x-amzn-transcribe-vocabulary-filter-name": self.request_shape.vocab_filter_name,
-            "x-amzn-transcribe-show-speaker-label": self.request_shape.show_speaker_label,
-            "x-amzn-transcribe-enable-channel-identification": self.request_shape.enable_channel_identification,
-            "x-amzn-transcribe-number-of-channels": self.request_shape.number_of_channels,
-        }
+    def _serialize_header(
+        self,
+        header: str,
+        value: Any,
+        expected_type: type,
+        prefix: str = "x-amzn-transcribe-",
+    ):
+        if value is None:
+            return {}
+        elif not isinstance(value, expected_type):
```

Comment:
I'd rather not add parameter validation as a part of this PR -- particularly not as a part of serialization. I think it makes more sense to have a discrete parameter validation step prior to serialization that can be more explicitly customized (e.g. disabled entirely if you just want to rely on `mypy`).